File: backend/Services/EmailAnalysisService.cs
using System.Text.RegularExpressions;
using CareerMail.Api.Data;
using CareerMail.Api.Models.Entities;
using CareerMail.Api.Models.Enums;
using CareerMail.Api.Services.Analyzer;
using Microsoft.EntityFrameworkCore;

namespace CareerMail.Api.Services;

public interface IEmailAnalysisService
{
    Task<EmailAnalysisService.ProcessOutcome> ProcessEmailAsync(Email email, User user);
    Task ReprocessAllUserEmailsAsync(User user);
}

public class EmailAnalysisService : IEmailAnalysisService
{
    private readonly IEmailAnalyzer _emailAnalyzer;
    private readonly AppDbContext _context;

    public EmailAnalysisService(IEmailAnalyzer emailAnalyzer, AppDbContext context)
    {
        _emailAnalyzer = emailAnalyzer;
        _context = context;
    }

    public record ProcessOutcome(
        bool IsJobRelated,
        bool ApplicationCreated,
        bool ApplicationUpdated,
        bool InterviewCreated,
        bool FollowUpCreated
    );

    public async Task<ProcessOutcome> ProcessEmailAsync(Email email, User user)
    {
        var result = _emailAnalyzer.Analyze(email.Subject, email.Body, email.Sender, email.SenderEmail);

        if (!result.JobRelated)
        {
            email.IsJobRelated = false;
            email.Classification = null;
            email.DetectedCompany = null;
            email.DetectedRole = null;
            email.DetectedStatus = null;
            email.ProcessedAt = DateTime.UtcNow;
            return new ProcessOutcome(false, false, false, false, false);
        }

        email.IsJobRelated = true;
        email.DetectedCompany = result.Company;
        email.DetectedRole = result.JobTitle;
        email.DetectedStatus = result.Status.HasValue ? result.Status.Value.ToString() : ApplicationStatus.APPLIED.ToString();
        email.DetectedRecruiterName = result.RecruiterName;
        email.DetectedRecruiterEmail = result.RecruiterEmail;
        email.DetectedRecruiterTitle = result.RecruiterTitle;
        email.DetectedRecruiterType = result.RecruiterType;
        email.DetectedRecruiterConfidence = result.ContactConfidence;
        email.Classification = result.Classification;
        email.ProcessedAt = DateTime.UtcNow;

        var company = !string.IsNullOrWhiteSpace(result.Company) ? result.Company : "Unknown Company";

        // Thread Matching first
        JobApplication? matchedApp = null;
        if (!string.IsNullOrWhiteSpace(email.GmailThreadId))
        {
            var threadEmail = await _context.Emails
                .Include(e => e.JobApplication)
                .FirstOrDefaultAsync(e => e.UserId == user.Id && e.GmailThreadId == email.GmailThreadId && e.JobApplicationId != null);
            matchedApp = threadEmail?.JobApplication;
        }

        // Company Name Matching second
        if (matchedApp == null && !string.Equals(company, "Unknown Company", StringComparison.OrdinalIgnoreCase))
        {
            matchedApp = await _context.JobApplications
                .FirstOrDefaultAsync(ja => ja.UserId == user.Id && EF.Functions.ILike(ja.Company, company));
        }

        if (matchedApp == null && result.Classification == EmailClassification.NEW_OPPORTUNITY)
        {
            return new ProcessOutcome(true, false, false, false, false);
        }

        JobApplication app;
        var appCreated = false;
        var appUpdated = false;
        var interviewCreated = false;
        var followUpCreated = false;

        if (matchedApp != null)
        {
            app = matchedApp;
            appUpdated = true;

            if (!string.IsNullOrWhiteSpace(company) && !company.Equals("Unknown Company", StringComparison.OrdinalIgnoreCase))
            {
                app.Company = company;
                app.CompanyLogo = Regex.Replace(company.ToLowerInvariant(), "[^a-z0-9]", "");
            }
            if (!string.IsNullOrWhiteSpace(result.JobTitle) && !result.JobTitle.Equals("Applicant", StringComparison.OrdinalIgnoreCase))
            {
                app.Title = result.JobTitle;
            }

            if (result.Status.HasValue && ShouldUpgradeStatus(app.Status, result.Status.Value))
            {
                app.Status = result.Status.Value;
            }

            if (result.Salary != null && app.Salary == null)
            {
                app.Salary = result.Salary;
            }

            if (!string.IsNullOrWhiteSpace(result.RecruiterName))
            {
                var shouldUpdate = app.RecruiterName == null
                    || (result.RecruiterType == RecruiterType.HUMAN_RECRUITER && app.RecruiterType != RecruiterType.HUMAN_RECRUITER)
                    || (result.ContactConfidence.HasValue && (!app.ContactConfidence.HasValue || result.ContactConfidence.Value > app.ContactConfidence.Value));

                if (shouldUpdate)
                {
                    app.RecruiterName = result.RecruiterName;
                    app.RecruiterEmail = result.RecruiterEmail;
                    app.RecruiterTitle = result.RecruiterTitle;
                    app.RecruiterPhone = result.RecruiterPhone;
                    app.RecruiterLinkedin = result.RecruiterLinkedin;
                    app.RecruiterType = result.RecruiterType;
                    app.ContactConfidence = result.ContactConfidence;
                    app.ContactExtractionSource = result.ContactExtractionSource;
                }
            }

            var emailDate = DateOnly.FromDateTime(email.Timestamp);
            if (!app.DateApplied.HasValue || emailDate < app.DateApplied.Value)
            {
                app.DateApplied = emailDate;
            }

            app.LastActivityDate = DateOnly.FromDateTime(DateTime.UtcNow);
            app.ActivitySubtitle = ComputeSubtitle(app.Status);

            app.TimelineEvents.Add(new TimelineEvent
            {
                Title = email.Subject,
                Description = result.TimelineNote ?? email.Preview ?? email.Subject,
                EventDate = email.Timestamp,
                EventType = result.Status?.ToString() ?? app.Status.ToString()
            });

            _context.JobApplications.Update(app);
        }
        else
        {
            appCreated = true;
            app = new JobApplication
            {
                UserId = user.Id,
                Company = company,
                Title = !string.IsNullOrWhiteSpace(result.JobTitle) ? result.JobTitle : "Applicant",
                Location = result.Location,
                EmploymentType = result.EmploymentType,
                Salary = result.Salary,
                DateApplied = DateOnly.FromDateTime(email.Timestamp),
                LastActivityDate = DateOnly.FromDateTime(DateTime.UtcNow),
                Status = result.Status ?? ApplicationStatus.APPLIED,
                Priority = ComputePriority(result.Status ?? ApplicationStatus.APPLIED),
                RecruiterName = result.RecruiterName,
                RecruiterEmail = result.RecruiterEmail,
                RecruiterTitle = result.RecruiterTitle,
                RecruiterPhone = result.RecruiterPhone,
                RecruiterLinkedin = result.RecruiterLinkedin,
                RecruiterType = result.RecruiterType,
                ContactConfidence = result.ContactConfidence,
                ContactExtractionSource = result.ContactExtractionSource,
                Source = "Gmail Auto-Detection",
                CompanyLogo = Regex.Replace(company.ToLowerInvariant(), "[^a-z0-9]", ""),
                ActivitySubtitle = ComputeSubtitle(result.Status ?? ApplicationStatus.APPLIED),
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            app.TimelineEvents.Add(new TimelineEvent
            {
                Title = $"Discovered: {app.Title} at {app.Company}",
                Description = result.TimelineNote ?? $"Detected from email: {email.Subject}",
                EventDate = email.Timestamp,
                EventType = app.Status.ToString()
            });

            _context.JobApplications.Add(app);
            await _context.SaveChangesAsync();
        }

        email.JobApplicationId = app.Id;

        // Auto-create Interview
        if (result.Classification == EmailClassification.INTERVIEW_INVITATION ||
            result.Classification == EmailClassification.INTERVIEW_SCHEDULED ||
            result.Status == ApplicationStatus.INTERVIEW ||
            result.Status == ApplicationStatus.FINAL_INTERVIEW)
        {
            var intDate = result.InterviewDateTime ?? DateTime.UtcNow.AddDays(3).Date.AddHours(11);
            var exists = await _context.Interviews.AnyAsync(i => i.UserId == user.Id && i.JobApplicationId == app.Id && i.InterviewDate == intDate);

            if (!exists)
            {
                var interview = new Interview
                {
                    UserId = user.Id,
                    JobApplicationId = app.Id,
                    Company = app.Company,
                    Title = app.Title,
                    InterviewDate = intDate,
                    Type = result.InterviewType ?? "Technical Interview",
                    Interviewer = result.RecruiterName ?? "Hiring Team",
                    Location = result.InterviewLink != null ? "Online / Video" : "Online",
                    MeetingLink = result.InterviewLink,
                    DaysAwayBadge = "In 3 days",
                    CompanyLogo = app.CompanyLogo,
                    PreparationNotes = $"Prepare technical background and experience for {app.Company}.",
                    Status = InterviewStatus.SCHEDULED
                };
                _context.Interviews.Add(interview);
                interviewCreated = true;
            }
        }

        // Auto-create Follow-up
        if (result.Classification == EmailClassification.ASSESSMENT || result.Deadline.HasValue)
        {
            var dueDate = result.Deadline ?? DateOnly.FromDateTime(DateTime.UtcNow.AddDays(5));
            var exists = await _context.FollowUps.AnyAsync(f => f.UserId == user.Id && f.JobApplicationId == app.Id && f.DueDate == dueDate);

            if (!exists)
            {
                var days = dueDate.DayNumber - DateOnly.FromDateTime(DateTime.UtcNow).DayNumber;
                var followUp = new FollowUp
                {
                    UserId = user.Id,
                    JobApplicationId = app.Id,
                    Company = app.Company,
                    Role = app.Title,
                    DueDate = dueDate,
                    AppliedSubtitle = "Assessment Active",
                    DaysDueBadge = $"Due in {days} days",
                    CompanyLogo = app.CompanyLogo,
                    Notes = $"Complete online coding assessment for {app.Company} before deadline.",
                    Status = FollowUpStatus.PENDING
                };
                _context.FollowUps.Add(followUp);
                followUpCreated = true;
            }
        }

        await _context.SaveChangesAsync();
        return new ProcessOutcome(true, appCreated, appUpdated, interviewCreated, followUpCreated);
    }

    public async Task ReprocessAllUserEmailsAsync(User user)
    {
        var interviews = await _context.Interviews.Where(i => i.UserId == user.Id).ToListAsync();
        _context.Interviews.RemoveRange(interviews);

        var followUps = await _context.FollowUps.Where(f => f.UserId == user.Id).ToListAsync();
        _context.FollowUps.RemoveRange(followUps);

        var userApps = await _context.JobApplications
            .Where(a => a.UserId == user.Id && (a.Source == "Gmail Auto-Detection" || a.Source == null))
            .ToListAsync();

        foreach (var app in userApps)
        {
            var linked = await _context.Emails.Where(e => e.UserId == user.Id && e.JobApplicationId == app.Id).ToListAsync();
            foreach (var e in linked)
            {
                e.JobApplicationId = null;
                e.IsJobRelated = false;
            }
            _context.JobApplications.Remove(app);
        }

        await _context.SaveChangesAsync();

        var allEmails = await _context.Emails.Where(e => e.UserId == user.Id).OrderBy(e => e.Timestamp).ToListAsync();
        foreach (var email in allEmails)
        {
            await ProcessEmailAsync(email, user);
        }
    }

    private static bool ShouldUpgradeStatus(ApplicationStatus current, ApplicationStatus incoming)
    {
        if (current == incoming) return false;
        if (current == ApplicationStatus.OFFER && incoming != ApplicationStatus.REJECTED && incoming != ApplicationStatus.WITHDRAWN)
        {
            return false;
        }
        return GetStatusWeight(incoming) >= GetStatusWeight(current);
    }

    private static int GetStatusWeight(ApplicationStatus status) => status switch
    {
        ApplicationStatus.APPLIED => 1,
        ApplicationStatus.ASSESSMENT => 2,
        ApplicationStatus.RECRUITER_SCREEN => 3,
        ApplicationStatus.INTERVIEW => 4,
        ApplicationStatus.FINAL_INTERVIEW => 5,
        ApplicationStatus.OFFER => 6,
        ApplicationStatus.REJECTED => 7,
        ApplicationStatus.WITHDRAWN => 8,
        _ => 1
    };

    private static Priority ComputePriority(ApplicationStatus status) =>
        status is ApplicationStatus.OFFER or ApplicationStatus.FINAL_INTERVIEW or ApplicationStatus.INTERVIEW
            ? Priority.HIGH
            : Priority.MEDIUM;

    private static string ComputeSubtitle(ApplicationStatus status) => status switch
    {
        ApplicationStatus.ASSESSMENT => "Assessment active",
        ApplicationStatus.RECRUITER_SCREEN => "Screening call",
        ApplicationStatus.INTERVIEW => "Technical Interview",
        ApplicationStatus.FINAL_INTERVIEW => "Final Round",
        ApplicationStatus.OFFER => "Offer Received 🎉",
        ApplicationStatus.REJECTED => "Application closed",
        ApplicationStatus.WITHDRAWN => "Application withdrawn",
        _ => "Applied recently"
    };
}

File: backend/app/services/email_analysis_service.py
import re
from datetime import datetime, date, timedelta
from sqlalchemy.orm import Session
from app.models.models import Email, JobApplication, TimelineEvent, Interview, FollowUp
from app.models.enums import ApplicationStatus, EmailClassification, InterviewStatus, FollowUpStatus, RecruiterType
from app.services.analyzer.rule_based_email_analyzer import RuleBasedEmailAnalyzer
from app.services.analyzer.analysis_result import EmailAnalysisResult


class EmailAnalysisService:
    def __init__(self, db: Session, analyzer: RuleBasedEmailAnalyzer | None = None):
        self.db = db
        self.analyzer = analyzer or RuleBasedEmailAnalyzer()

    def process_and_link_email(self, email: Email, user_id: int) -> EmailAnalysisResult:
        result = self.analyzer.analyze(
            sender=email.sender,
            sender_email=email.sender_email,
            subject=email.subject,
            body=email.body
        )

        email.is_job_related = result.is_job_related
        if not result.is_job_related:
            email.processed_at = datetime.utcnow()
            return result

        email.classification = result.classification.value if result.classification else None
        email.detected_company = result.detected_company
        email.detected_role = result.detected_role
        email.detected_status = result.suggested_status.value if result.suggested_status else None

        if result.recruiter_info:
            email.detected_recruiter_name = result.recruiter_info.name
            email.detected_recruiter_email = result.recruiter_info.email
            email.detected_recruiter_title = result.recruiter_info.title
            email.detected_recruiter_type = result.recruiter_info.recruiter_type.value
            email.detected_recruiter_confidence = result.recruiter_info.confidence

        email.processed_at = datetime.utcnow()

        # Try to find or link existing JobApplication
        app = None
        if result.detected_company:
            comp_clean = result.detected_company.strip().lower()
            app = (
                self.db.query(JobApplication)
                .filter(
                    JobApplication.user_id == user_id,
                    JobApplication.company.ilike(f"%{comp_clean}%")
                )
                .first()
            )

        if not app and result.detected_company and result.classification != EmailClassification.REJECTION:
            # Auto-create new application
            logo = re.sub(r"[^a-z0-9]", "", result.detected_company.lower())
            app = JobApplication(
                user_id=user_id,
                company=result.detected_company,
                title=result.detected_role or "Software Engineer",
                location="Remote / Hybrid",
                employment_type="Full-time",
                date_applied=date.today(),
                status=result.suggested_status.value if result.suggested_status else ApplicationStatus.APPLIED.value,
                company_logo=logo,
                source="Email Inbound",
                recruiter_name=result.recruiter_info.name,
                recruiter_email=result.recruiter_info.email,
                recruiter_title=result.recruiter_info.title,
                recruiter_phone=result.recruiter_info.phone,
                recruiter_linkedin=result.recruiter_info.linkedin_url,
                recruiter_type=result.recruiter_info.recruiter_type.value,
                contact_confidence=result.recruiter_info.confidence,
                contact_extraction_source=result.recruiter_info.extraction_source,
                last_activity_date=date.today(),
                activity_subtitle=f"Inbound from {result.detected_company}",
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )
            app.timeline_events.append(TimelineEvent(
                title=f"Application Initialized: {app.title}",
                description=f"Auto-created from email inbound: '{email.subject}'",
                event_date=datetime.utcnow(),
                event_type="EMAIL_INBOUND"
            ))
            self.db.add(app)
            self.db.commit()
            self.db.refresh(app)

        if app:
            email.job_application_id = app.id

            # Apply Status Upgrade if higher or Rejection/Offer
            if result.suggested_status:
                self._update_application_status(app, result.suggested_status, result.timeline_title, result.timeline_description)

            # Update Recruiter if higher confidence
            if result.recruiter_info and result.recruiter_info.confidence > (app.contact_confidence or 0):
                if result.recruiter_info.name:
                    app.recruiter_name = result.recruiter_info.name
                if result.recruiter_info.email:
                    app.recruiter_email = result.recruiter_info.email
                if result.recruiter_info.title:
                    app.recruiter_title = result.recruiter_info.title
                if result.recruiter_info.phone:
                    app.recruiter_phone = result.recruiter_info.phone
                if result.recruiter_info.linkedin_url:
                    app.recruiter_linkedin = result.recruiter_info.linkedin_url
                app.recruiter_type = result.recruiter_info.recruiter_type.value
                app.contact_confidence = result.recruiter_info.confidence
                app.contact_extraction_source = result.recruiter_info.extraction_source

            # Auto-schedule Interview if detected
            if result.classification == EmailClassification.INTERVIEW_INVITATION:
                int_date = result.interview_date or (datetime.utcnow() + timedelta(days=2, hours=3))
                existing_int = (
                    self.db.query(Interview)
                    .filter(Interview.job_application_id == app.id, Interview.user_id == user_id)
                    .first()
                )
                if not existing_int:
                    intv = Interview(
                        user_id=user_id,
                        job_application_id=app.id,
                        company=app.company,
                        title=f"{result.interview_type or 'Technical Interview'} - {app.company}",
                        interview_date=int_date,
                        type=result.interview_type or "Technical Interview",
                        interviewer=result.recruiter_info.name or "Hiring Team",
                        location="Google Meet",
                        meeting_link=result.meeting_link or "https://meet.google.com/new",
                        status=InterviewStatus.SCHEDULED.value,
                        company_logo=app.company_logo
                    )
                    self.db.add(intv)

            # Auto-schedule FollowUp if assessment
            if result.classification == EmailClassification.ASSESSMENT:
                existing_fol = (
                    self.db.query(FollowUp)
                    .filter(FollowUp.job_application_id == app.id, FollowUp.user_id == user_id)
                    .first()
                )
                if not existing_fol:
                    fol = FollowUp(
                        user_id=user_id,
                        job_application_id=app.id,
                        company=app.company,
                        role=app.title,
                        due_date=date.today() + timedelta(days=3),
                        applied_subtitle="Assessment Active",
                        status=FollowUpStatus.PENDING.value,
                        company_logo=app.company_logo,
                        notes="Complete and submit online assessment."
                    )
                    self.db.add(fol)

            app.last_activity_date = date.today()
            app.updated_at = datetime.utcnow()
            self.db.commit()

        return result

    def _update_application_status(
        self,
        app: JobApplication,
        new_status: ApplicationStatus,
        title: str | None,
        description: str | None
    ):
        status_rank = {
            ApplicationStatus.APPLIED: 1,
            ApplicationStatus.ASSESSMENT: 2,
            ApplicationStatus.RECRUITER_SCREEN: 3,
            ApplicationStatus.INTERVIEW: 4,
            ApplicationStatus.FINAL_INTERVIEW: 5,
            ApplicationStatus.OFFER: 6,
            ApplicationStatus.REJECTED: 7,
            ApplicationStatus.WITHDRAWN: 8,
        }

        current_rank = status_rank.get(ApplicationStatus.from_string(app.status), 1)
        target_rank = status_rank.get(new_status, 1)

        # Allow upgrade if higher rank or offer/rejection
        if target_rank > current_rank or new_status in (ApplicationStatus.OFFER, ApplicationStatus.REJECTED):
            app.status = new_status.value
            subtitle_map = {
                ApplicationStatus.ASSESSMENT: "Assessment invited",
                ApplicationStatus.RECRUITER_SCREEN: "Screening call",
                ApplicationStatus.INTERVIEW: "Technical Interview",
                ApplicationStatus.FINAL_INTERVIEW: "Final Round",
                ApplicationStatus.OFFER: "Offer Received 🎉",
                ApplicationStatus.REJECTED: "Application closed",
            }
            app.activity_subtitle = subtitle_map.get(new_status, "Status updated")

            app.timeline_events.append(TimelineEvent(
                title=title or f"Moved to {new_status.get_display_name()}",
                description=description or f"Updated via incoming email",
                event_date=datetime.utcnow(),
                event_type=new_status.value
            ))
